File: android/src/main/java/com/example/android/app/ui/search/SearchResultUiModel.kt
package com.example.android.app.ui.search

import com.example.shared.data.source.local.db.Repo
import com.example.shared.model.PagedRepoSearchResult

data class SearchResultUiModel(
    val inProgress: Boolean = false,
    val searchResult: PagedRepoSearchResult? = null,
    val error: String? = null
) {
    val hasNoState: Boolean
        get() = !inProgress && searchResult == null && error == null

    val firstInProgress: Boolean
        get() = inProgress && searchResult == null

    val pagingInProgress: Boolean
        get() = inProgress && searchResult != null

    val visibleSearchResult: Boolean
        get() = !searchResult?.repos.isNullOrEmpty()

    val visibleEmptySearchResult: Boolean
        get() = hasNoState || searchResult?.repos?.isEmpty() == true
}

fun PagedRepoSearchResult.merge(existingRepos: List<Repo>) =
    copy(existingRepos + repos)

File: ios/Sources/UI/Search/SearchResultUiModel.swift
import Foundation
import Shared

struct SearchResultUiModel {
    let inProgress: Bool
    let searchResult: PagedRepoSearchResult?
    let error: String?

    var hasNoState: Bool {
        !inProgress && searchResult == nil && error == nil
    }

    var firstInProgress: Bool {
        inProgress && searchResult == nil
    }

    var pagingInProgress: Bool {
        inProgress && searchResult != nil
    }

    init(inProgress: Bool = false,
         searchResult: PagedRepoSearchResult? = nil,
         error: String? = nil)
    {
        self.inProgress = inProgress
        self.searchResult = searchResult
        self.error = error
    }
}
